src/records.rs

```diff
@@ -0,0 +1,303 @@
+        loader: &mut AtomicStoreLoader,
+        adaptor: EncryptingResourceAdapter<(u64, Option<Record>)>,
+        fill_size: u64,
+    ) -> Result<Self, KeystoreError<L>> {
+        let log = AppendLog::load(loader, adaptor, "keystore_records", fill_size)?;
+        let store = RecordsStore::new(log)?;
+        let mut records = Self {
+            store,
+            asset_records: Persistable::new(),
+            nullifier_records: Persistable::new(),
+        };
+        records.reload();
+        Ok(records)
+    }
+
+    /// Create a Record
+    ///
+    /// Returns an editor to the newly created Record
+    pub fn create<L: Ledger>(
+        &mut self,
+        uid: u64,
+        ro: RecordOpening,
+        nullifier: Nullifier,
+        hold_until: Option<u64>,
+    ) -> Result<RecordEditor<'_>, KeystoreError<L>> {
+        let record = Record {
+            ro,
+            uid,
+            nullifier,
+            hold_until,
+        };
+        let ro = &record.ro;
+        self.asset_records.insert((
+            (ro.asset_def.code, ro.pub_key.address(), ro.freeze_flag),
+            (record.ro.amount.into(), record.uid),
+        ));
+        self.nullifier_records
+            .insert((record.nullifier, record.uid));
+        let mut editor = RecordEditor::new(self, record);
+        editor.save()?;
+        Ok(editor)
+    }
+
+    /// Rebuild the indices for this store
+    fn reload(&mut self) {
+        self.asset_records = Persistable::new();
+        self.nullifier_records = Persistable::new();
+        for record in self.store.iter() {
+            let ro = &record.ro;
+            self.asset_records.insert((
+                (
+                    ro.asset_def.code,
+                    ro.pub_key.address().clone(),
+                    ro.freeze_flag,
+                ),
+                (record.ro.amount.into(), record.uid),
+            ));
+            self.nullifier_records
+                .insert((record.nullifier, record.uid));
+        }
+    }
+
+    fn store<L: Ledger>(&mut self, uid: u64, record: &Record) -> Result<(), KeystoreError<L>> {
+        self.store.store(&uid, record)?;
+        self.reload();
+        Ok(())
+    }
+
+    /// Returns an Iterator to all the Records in the store
+    pub fn iter(&self) -> impl Iterator<Item = Record> + '_ {
+        self.store.iter().cloned()
+    }
+
+    /// Get a record from a uid from the store
+    pub fn get<L: Ledger>(&self, uid: u64) -> Result<Record, KeystoreError<L>> {
+        Ok(self.store.load(&uid)?)
+    }
+
+    /// get a Record Editor by uid from storage
+    pub fn get_mut<L: Ledger>(&mut self, uid: u64) -> Result<RecordEditor, KeystoreError<L>> {
+        let record = self.store.load(&uid)?;
+        Ok(RecordEditor::new(self, record))
+    }
+
+    /// Get records associated with an asset and account which are either frozen or unfrozen
+    /// Useful for finding records for transaction inputs
+    pub fn with_asset<L: Ledger>(
```

Comment:
`with_asset` feels like the wrong name for this method, since it's actually keyed on 3 fields, not just `asset`. Not sure what's better. If we break the pattern a little bit, we could call this `spendable` or something which sort of reflects the purpose of the index.